### PR TITLE
Convert parameters in CloudBigtableConfiguration to runtime parameters

### PR DESCRIPTION
**Pull request by kevinsi4508**
Though I tried to have minimum change, 'request' object in CloudBigtableScanConfiguration needs to be converted to runtime parameter as well because it depends on project ID and instance ID.

### CODE REVIEW

**Comment by kevinsi4508** [comment]
I still need to update the ordering of those .with* methods.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 1]
```diff
@@ -64,11 +67,11 @@ public static CloudBigtableScanConfiguration fromConfig(CloudBigtableTableConfig
     private Scan scan;
     private ReadRowsRequest request;
 
-    public Builder() {
-    }
+    public Builder() {}
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: please revert the formatting.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 2]
```diff
@@ -43,15 +46,15 @@
   private static final long serialVersionUID = 2435897354284600685L;
 
   /**
-   * Converts a {@link CloudBigtableTableConfiguration} object to a
-   * {@link CloudBigtableScanConfiguration} that will perform the specified {@link Scan} on the
-   * table.
+   * Converts a {@link CloudBigtableTableConfiguration} object to a {@link
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: please revert the formatting

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 3]
```diff
@@ -43,15 +46,15 @@
...
+   * CloudBigtableScanConfiguration} that will perform the specified {@link Scan} on the table.
+   *
    * @param config The {@link CloudBigtableTableConfiguration} object.
    * @param scan The {@link Scan} to add to the configuration.
    * @return The new {@link CloudBigtableScanConfiguration}.
    */
-  public static CloudBigtableScanConfiguration fromConfig(CloudBigtableTableConfiguration config,
-      Scan scan) {
+  public static CloudBigtableScanConfiguration fromConfig(
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: please revert the formatting

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 4]
```diff
@@ -117,6 +123,9 @@ public Builder withProjectId(String projectId) {
 
     /**
      * {@inheritDoc}
+     *
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: please move these comment to a different PR.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 5]
```diff
@@ -231,14 +262,14 @@ ByteString getStopRowByteString() {
   }
 
   RowRange getRowRange() {
-    RowSet rows = request.getRows();
+    RowSet rows = getRequest().getRows();
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: nit, please move the encapsulation to a different PR

Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: It is either "getRequest()" or "request.get()". Since we are modifying the code anyway, I would rather do this now.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 1 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: done.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 2 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: done.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 3 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: done.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 4 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: done.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 6]
```diff
@@ -173,25 +182,38 @@ public CloudBigtableScanConfiguration build() {
...
+                if (!request.getTableName().isEmpty()) {
+                  return request;
+                }
+
+                BigtableInstanceName bigtableInstanceName =
+                    new BigtableInstanceName(getProjectId(), getInstanceId());
+                String fullTableName = bigtableInstanceName.toTableNameStr(getTableId());
+                return request.toBuilder().setTableName(fullTableName).build();
+              }
+            });
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: I dont think this is right...`this.request` will end up with an incomplete `ValueProvider` that will return `true` for `isAccessible` when projectId or instanceId are not accessible. I think you need to create a your own subclass of `ValueProvider` that wraps all of the relevant `ValueProvider`s and in its implementation `isAccessible` it takes an && of all its component ValueProviders.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 7]
```diff
@@ -265,8 +287,9 @@ public ByteKeyRange toByteKeyRange() {
   @Override
   public void populateDisplayData(DisplayData.Builder builder) {
     super.populateDisplayData(builder);
-    builder
-        .add(DisplayData.item("readRowsRequest", request.toString()).withLabel("ReadRowsRequest"));
+    if (areParametersAccessible()) {
+      builder.add(
+          DisplayData.item("readRowsRequest", request.toString()).withLabel("ReadRowsRequest"));
+    }
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: To ease debugging, maybe set the value to "Unavailable during pipeline construction" or something if the params are not accessible. It will give users something to copy/paste into a bug report

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 6 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: I don't think this matters at the moment, no? If you use CloudBigtableScanConfiguration now, there is no way you can set runtime parameters, therefore isAccessible is a non-issue.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 7 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: I don't think user will see this at all at construction time.
I have confirmed this actually showed up correctly at runtime.

Comment (igorbernstein2) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: I didn't mena for "Unavailable during pipeline construction"  to show up in the normal case. I meant for it show up in case something does wrong and we need to debug. It will be easier for the user to describe this message rather than the absence of it.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 6 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: The request object can't be modified before wrapping it in StaticValueProvider. I added TODO.

Comment (igorbernstein2) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: SGTM

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 7 above)
Comment (kevinsi4508) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: Added TODO.
